app/services/notifications_service.rb

```diff
@@ -0,0 +1,60 @@
+class NotificationsService
+
+  @@gcm = GCM.new(ENV['GOOGLE_API_KEY'])
+
+  def self.notify_users_on_trip(trip)
+    users = []
+    if trip.shopper.tripping?
+      users = User.idle.where.not(notification_key: nil)
+    elsif trip.shopper.idle?
+      # meaning trip was cancelled
+      # TODO: complete this
+    end
+
+    Rails.logger.debug "Sending notification to #{users.length} users"
+    users.each do |requester|
+      response = @@gcm.send_with_notification_key(requester.notification_key, data: TripNotificationSerializer.new(trip).as_json, collapse_key: "trip_#{trip.id}")
+      if response[:response] == 'success'
+        Rails.logger.debug "Successfully sent notification to requester id #{requester.id} on trip with id: #{trip.id}"
+      else
+        Rails.logger.error "Failed to send notification to requester id #{requester.id} on trip with id: #{trip.id}, due to #{response[:body]}"
+      end
+    end
+  end
+
+  def self.notify_shopper_on_request(request)
+    response = @@gcm.send_with_notification_key(request.trip.shopper.notification_key, data: ShopRequestNotificationSerializer.new(request).as_json)
```

Comment:
well, the only reason you would want a collapse key here is to if the user later cancels the request, but even then you would probably not want to replace the previous notification with a "<user> has cancelled his request", but rather you would want to delete the notification from the action bar altogether.